test/clj_gatling/simulation_test.clj

```diff
@@ -535,6 +535,96 @@
       (is #{1.0} @progress-distribution)
       (is (= (sort @progress-distribution) @progress-distribution)))))
 
+(deftest with-3-arity-concurrency-function
+  (let [concurrency-function-called? (atom false)
+        context-to-fn (atom {})
+        duration-distribution (atom [])]
+    (run-single-scenario {:name "scenario"
+                          :steps [(step "step" true)]}
+                         :concurrency 10
+                         :requests 100
+                         :context {:value 1}
+                         :concurrency-distribution (fn [progress duration context]
+                                                     (reset! context-to-fn context)
+                                                     (reset! concurrency-function-called? true)
+                                                     (swap! duration-distribution conj (.toMillis duration))
+                                                     (if (< progress 0.5)
+                                                       0.1
+                                                       1.0)))
+    (testing "concurrency-function is called"
+      (is (= true @concurrency-function-called?)))
+    (testing "context is passed to concurrency-function"
+      (is (= {:value 1} @context-to-fn)))
+    (testing "Duration has ordered values"
+      (is (> (count @duration-distribution) 10))
+      (is (= (sort @duration-distribution) @duration-distribution)))))
```

Comment:
Clock skew and contention over the atom does mean that this test occasionally fails because the timestamps recorded for the runs are occasionally a little out of order. I didn’t worry _too_ much about that as there were other slightly flaky tests already in the codebase, but some sort of `approximately==` for collections could be written if necessary.

Reply:
Yep. That is fine for now. If it turns out that these tests fail often I can take a look later on.